cmd/nvidia-cdi-hook/cudacompat/cudacompat.go

```diff
@@ -0,0 +1,247 @@
+	}
+	if !containerRoot.hasPath(cudaCompatPath) {
+		return nil
+	}
+	if !containerRoot.hasPath("/etc/ld.so.cache") {
+		// If there is no ldcache in the container, the hook is a no-op.
+		return nil
+	}
+
+	libs, err := containerRoot.globFiles(filepath.Join(cudaCompatPath, "libcuda.so.*.*"))
+	if err != nil {
+		m.logger.Warningf("Failed to find CUDA compat library: %w", err)
+		return nil
+	}
+
+	if len(libs) == 0 {
+		return nil
+	}
+
+	if len(libs) != 1 {
+		m.logger.Warningf("Unexpected number of CUDA compat libraries: %v", libs)
+		return nil
+	}
+
+	compatDriverVersion := strings.TrimPrefix(filepath.Base(libs[0]), "libcuda.so.")
+	compatMajor := strings.SplitN(compatDriverVersion, ".", 2)[0]
+
+	driverMajor := strings.SplitN(hostDriverVersion, ".", 2)[0]
+
+	if driverMajor < compatMajor {
+		resolvedCompatDir := strings.TrimPrefix(filepath.Dir(libs[0]), string(containerRoot))
+		return m.createLdsoconfdFile(containerRoot, cudaCompatLdsoconfdFilenamePattern, resolvedCompatDir)
+	}
+	return nil
+}
+
+// A root is used to add basic path functionality to a string.
+type root string
+
+// hasPath checks whether the specified path exists in the root.
+func (r root) hasPath(path string) bool {
+	resolved, err := r.resolve(path)
+	if err != nil {
+		return false
+	}
+	if _, err := os.Stat(resolved); err != nil && os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// globFiles matches the specified pattern in the root.
+// The files that match must be regular files.
+func (r root) globFiles(pattern string) ([]string, error) {
+	patternPath, err := r.resolve(pattern)
+	if err != nil {
+		return nil, err
+	}
+	matches, err := filepath.Glob(patternPath)
+	if err != nil {
+		return nil, err
+	}
+	var files []string
+	for _, match := range matches {
+		info, err := os.Lstat(match)
+		if err != nil {
+			return nil, err
+		}
+		// Ignore symlinks.
+		if info.Mode()&os.ModeSymlink != 0 {
+			continue
+		}
+		// Ignore directories.
+		if info.IsDir() {
+			continue
+		}
+		files = append(files, match)
+	}
+	return files, nil
+}
+
+// resolve returns the absolute path including root path.
+// Symlinks are resolved, but are guaranteed to resolve in the root.
+func (r root) resolve(path string) (string, error) {
+	absolute := filepath.Clean(filepath.Join(string(r), path))
+	return symlink.FollowSymlinkInScope(absolute, string(r))
+}
+
+// createLdsoconfdFile creates a file at /etc/ld.so.conf.d/ in the specified root.
+// The file is created at /etc/ld.so.conf.d/{{ .pattern }} using `CreateTemp` and
+// contains the specified folders on each line.
+func (m command) createLdsoconfdFile(in root, pattern string, folders ...string) error {
+	if len(folders) == 0 {
+		m.logger.Debugf("No folders to add to /etc/ld.so.conf")
+		return nil
+	}
+
+	ldsoconfdDir, err := in.resolve("/etc/ld.so.conf.d")
+	if err != nil {
+		return err
+	}
+	if err := os.MkdirAll(ldsoconfdDir, 0755); err != nil {
+		return fmt.Errorf("failed to create ld.so.conf.d: %w", err)
+	}
```

Comment:
What is happening here? You try and resolve `ldsoconfdDir` and only if there is not error, you attempt to create it?

Reply:
`containerRoot.resolve` resolves a path in the container root -- following any symlinks and ensuring that we do not escape the container filesystem. The final path does not have to exsit which is why we create it. (`MkdirAll` is a no-op if the directory exists).